src/native/minipal/log.c

```diff
@@ -0,0 +1,329 @@
+}
+
+void minipal_log_flush(minipal_log_flags flags)
+{
+    ;
+}
+
+static size_t log_write_line(minipal_log_flags flags, const char* msg, size_t msg_len)
+{
+    char buffer[MINIPAL_LOG_MAX_PAYLOAD];
+    if (msg_len < MINIPAL_LOG_MAX_PAYLOAD)
+    {
+        strncpy(buffer, msg, msg_len);
+        buffer[msg_len] = '\0';
+        return log_write(flags, buffer, msg_len);
+    }
+
+    const char* msg_end = msg + msg_len;
+    size_t bytes_written = 0;
+    while (msg < msg_end)
+    {
+        ptrdiff_t chunk_size = MINIPAL_LOG_MAX_PAYLOAD - 1;
+        if (msg_end - msg < chunk_size)
+        {
+            chunk_size = msg_end - msg;
+        }
+
+        strncpy(buffer, msg, chunk_size);
+        buffer[chunk_size] = '\0';
+        bytes_written += log_write(flags, buffer, chunk_size);
+        msg += chunk_size;
+    }
+
+    return bytes_written;
+}
+
+int minipal_log_write(minipal_log_flags flags, const char* msg)
+{
+    if (msg == NULL || msg[0] == '\0')
+        return 0;
+
+    size_t msg_len = strlen(msg);
+    const char* msg_end = msg + msg_len;
+
+    if (msg_len < MINIPAL_LOG_MAX_PAYLOAD)
+        return (int)log_write(flags, msg, msg_len);
+
+    const char* next_msg = NULL;
+    size_t bytes_written = 0;
+    for (next_msg = msg; next_msg < msg_end;)
+    {
+        const char* next_line_break = strchr(next_msg, '\n');
+        if (next_line_break == NULL && (msg_end - next_msg < MINIPAL_LOG_MAX_PAYLOAD))
+        {
+            bytes_written += log_write(flags, next_msg, msg_end - next_msg);
+            break;
+        }
+        else if (next_line_break == NULL)
+        {
+            bytes_written += log_write_line(flags, next_msg, msg_end - next_msg);
+            break;
+        }
+        else
+        {
+            bytes_written += log_write_line(flags, next_msg, next_line_break - next_msg);
+            next_msg = next_line_break + 1;
+        }
+    }
+
+    return (int)bytes_written;
+}
+
+void minipal_log_sync(minipal_log_flags flags)
+{
+    ;
+}
+#else
+#include <stdio.h>
+
+#define MINIPAL_LOG_MAX_PAYLOAD 32767
+
+static FILE * get_std_file(minipal_log_flags flags)
+{
+    switch(flags)
+    {
+    case minipal_log_flags_fatal:
+    case minipal_log_flags_error:
+        return stderr;
+    case minipal_log_flags_warning:
+    case minipal_log_flags_info:
+    case minipal_log_flags_debug:
+    case minipal_log_flags_verbose:
+    default:
+        return stdout;
+    }
+}
+
+int minipal_log_print(minipal_log_flags flags, const char* fmt, ... )
+{
+    va_list args;
+    va_start(args, fmt);
+    int status = vfprintf(get_std_file(flags), fmt, args);
+    va_end(args);
+    return status;
+}
+
+int minipal_log_vprint(minipal_log_flags flags, const char* fmt, va_list args)
+{
+    return vfprintf(get_std_file(flags), fmt, args);
+}
+
+void minipal_log_flush(minipal_log_flags flags)
+{
+    FILE* file = get_std_file(flags);
+    if (file != NULL)
+        fflush(file);
+}
+
+#if defined(HOST_WINDOWS)
```

Comment:
Is there a benefit in implementing the windows code here by the means of the Windows APIs instead of using the same code as we use for linux/macos/other unixes except android?

Reply:
This is the implementation we end up with going through the PAL, I guess we could use _write instead of WriteFile and _commit instead of fsync/sync/fcntl for flushing files, but in the past, I normally used the Win32 API's for this WriteFile/FlushFileBuffers on Windows.

Reply:
The point is that with the long term goal of getting rid of the CoreCLR, not using the Windows APIs in new code wherever possible is a benefit.

Reply:
Looks like _commit won't work with stdout/stderr, so at least for low level IO sync we would need to use WIN32 FlushFileBuffers.